### PR TITLE
Allow inferring divisibility on unbacked SymInts and do replacement trick

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113165

We want something like torch.empty(i0, 12).view(4, -1, 12) to work.  Right now, it chokes on guards on data dependent accesses. It turns out we are very close to having it work based on experiments in https://github.com/pytorch/pytorch/issues/112347 if we do the replacement trick, setting i0 = i1 * 4 to explicitly encode in the divisibility; this is good enough for Sympy to be able to handle the rest.

There are two parts to this PR.

* First, we must discover that there is this divisibility constraint. The place where this happens on view is in `infer_size`; however, we are unable to discover the modulus test with `expect_true` because the condition is currently written with a Python boolean operator that forces guarding too early: `numel == newsize or (dim is not None and newsize > 0 and numel % newsize == 0)`. We rewrite this into an equivalent version which tests on dim being None or not first, before performing individual tests. The main nontrivial reasoning here is that I must show that my set of tests in the `dim is None` branch are sufficient when `numel == newsize`. However, if `numel == newsize`, then the modulus must pass. Thus this is equivalent.
* Given the modifications to `infer_size`, this suffices to produce a runtime assert `Eq(Mod(192*i0, 2304), 0)`. Now we must simply turn this into the replacement automatically. I wasn't really sure how to use Sympy to do this for me, so I just manually pattern matched for this particular expression form, and if it exists do the replacements.

Note that this is kind of only useful for export, because inductor chokes on views involving unbacked SymInts. That will be follow up.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Comment by ezyang** [comment]
cc @asmeurer

**Comment by ezyang** [comment]
@Chillee you added the original divisibility support for backed symints, I'm kind of wondering if we should do the replacement trick there, it might make it easier to produce better guard code.

File: torch/_prims_common/__init__.py [hunk 1]
```diff
@@ -844,22 +844,36 @@ def infer_size(shape: ShapeType, numel: int) -> Tuple[int, ...]:
...
-    )
-    if dim is not None:
-        # Convert to list to produce a compatible error message with core
-        # PyTorch, which prints sequences in square brackets.
-        shape = list(shape)
+    if dim is None:
+        torch._check(numel == newsize)
+    else:
         torch._check(
             newsize != 0,
```
Comment (nunoplopes) [inline torch/_prims_common/__init__.py:851]: this check is made redundant with the new one below.

**Comment by nunoplopes** [comment]
The logic looks sound to me! 👍

File: torch/_prims_common/__init__.py (hunk 1 above)
Comment (ezyang) [reply torch/_prims_common/__init__.py:851]: I know, but this error message is better :)

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -3129,6 +3129,33 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
...
+                    # torch.empty(i0, q).view(c, -1, q) work out
+                    p, q = mod_expr.args
+                    if isinstance(q, sympy.Number) and isinstance(p, sympy.Mul) and len(p.args) == 2:
+                        c, i0 = p.args
+                        # Given Mod(c * i0, q) == 0
+                        if (
+                            isinstance(c, sympy.Number) and
+                            isinstance(i0, sympy.Symbol) and
+                            self.is_unbacked_symint(i0) and
+                            int(q) % int(c) == 0
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:3142]: As discussed, you can get rid of this condition replacing `c` by `gcd(q, c)` as the smallest positive solution of `Mod(c * i0, q) == 0` is `i0 = q/gcd(q, c)`

**Comment by asmeurer** [comment]
I wonder if it makes sense for you to try to integrate this into your `try_solve` helper instead. What you are doing here is also solving, i.e., solving $192 i_{0} \bmod 2304 = 0$. The principle issue is that the "solution", $i_0 = 12i_1$ involves a new parameter, as is typical for solutions of Diophantine equations.

**Review by lezcano — APPROVED** [review]

**Review by aakhundov — APPROVED** [review]

**Comment by ezyang** [comment]
@asmeurer I think I'm amenable to factoring this out into its own function, but it's difficult to see how to integrate it into try_solve, since, as you say, it requires creating a new parameter.

**Pull request closed without merging**